File: js/queue/linkedListQueue.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class LinkedListQueue {
  constructor() {
    this.dummyHead = new Node(null);
    this.tail = this.dummyHead;
    this.size = 0;
  }

  getSize() {
    return this.size;
  }

  isEmpty() {
    return this.size === 0;
  }

  enqueue(elem) {
    const node = new Node(elem);
    this.tail.next = node;
    this.tail = node;
    this.size++;
  }

  dequeue() {
    if (this.tail === this.dummyHead) {
      throw new Error("Queue is empty!");
    }

    // console.log(this.dummyHead);
    const node = this.dummyHead.next;
    this.dummyHead.next = node.next;
    node.next = null;
    this.size--;
    return node.data;
  }

  getFront() {
    if (this.tail === this.dummyHead) {
      throw new Error("Queue is empty!");
    }

    return this.dummyHead.next.data;
  }

  toString() {
    let str = "LinkedListQueue: Front [ ";
    let current = this.dummyHead.next;
    while(current !== null) {
      str += current.data;
      if (current.next !== null) {
        str += ", "
      }
      current = current.next;
    }
    str += " ] Tail";
    return str;
  }
}

module.exports = LinkedListQueue;

File: go/linkedlistqueue/linked_list_queue.go
package linkedlistqueue

import (
	"fmt"
	"errors"
)

type LinkedListQueue struct {
	head *Node
	tail *Node
	size int
}

func New() *LinkedListQueue {
	return &LinkedListQueue{}
}

func (q *LinkedListQueue) GetSize() int {
	return q.size
}

func (q *LinkedListQueue) IsEmpty() bool {
	return q.size == 0
}

func (q *LinkedListQueue) Enqueue(elem interface{}) {
	node := &Node{data: elem,}

	if q.size == 0 {
		q.head = node
		q.tail = node
	} else {
		q.tail.next = node
		q.tail = node
	}
	q.size++
}

func (q *LinkedListQueue) Dequeue() (elem interface{}, err error) {
	if q.size == 0 {
		err = errors.New("Queue is Empty")
		return
	}

	node := q.head
	elem = node.data
	q.head = q.head.next
	node = nil
	q.size--
	return
}

func (q *LinkedListQueue) GetFront() (elem interface{}, err error) {
	if q.size == 0 {
		err = errors.New("Queue is Empty")
		return
	}

	elem = q.head.data
	return
}

func (q *LinkedListQueue) String() string {
	str := fmt.Sprintf("Linked List Queue: Size = %d\n", q.size)
	str += "Head [ "
	
	for current := q.head; current.next != nil; current = current.next {
		str += fmt.Sprint(current.data) + ", "
	}

	str += fmt.Sprint(q.tail.data) + " ] Tail "
	return str
}